src/Microsoft.AspNetCore.Mvc.Razor/RazorPageActivator.cs

```diff
@@ -70,159 +57,23 @@ public void Activate(IRazorPage page, ViewContext context)
-            Debug.Assert(matchingConstructor != null);
-
-            var parameters = new ParameterExpression[] { Expression.Parameter(parameterTypes[0]) };
-            var newExpression = Expression.New(matchingConstructor, parameters);
-            var castNewCall = Expression.Convert(
-                newExpression,
-                typeof(ViewDataDictionary));
-            var lambda = Expression.Lambda<CreateViewDataNested>(castNewCall, parameters);
-            return lambda.Compile();
-        }
-
-        private CreateViewDataRoot GetCreateViewDataRoot(Type viewDataDictionaryType)
-        {
-            var parameterTypes = new Type[] {
-                typeof(IModelMetadataProvider),
-                typeof(ModelStateDictionary) };
-            var matchingConstructor = viewDataDictionaryType.GetConstructor(parameterTypes);
-            Debug.Assert(matchingConstructor != null);
-
-            var parameters = new ParameterExpression[] {
-                Expression.Parameter(parameterTypes[0]),
-                Expression.Parameter(parameterTypes[1]) };
-            var newExpression = Expression.New(matchingConstructor, parameters);
-            var castNewCall = Expression.Convert(
-                newExpression,
-                typeof(ViewDataDictionary));
-            var lambda = Expression.Lambda<CreateViewDataRoot>(castNewCall, parameters);
-            return lambda.Compile();
-        }
-
-
-        private PropertyActivator<ViewContext> CreateActivateInfo(PropertyInfo property)
-        {
-            Func<ViewContext, object> valueAccessor;
-            if (typeof(ViewDataDictionary).IsAssignableFrom(property.PropertyType))
-            {
-                // Logic looks reversed in condition above but is OK. Support only properties of base
-                // ViewDataDictionary type and activationInfo.ViewDataDictionaryType. VDD<AnotherType> will fail when
-                // assigning to the property (InvalidCastException) and that's fine.
-                valueAccessor = context => context.ViewData;
-            }
-            else if (property.PropertyType == typeof(IUrlHelper))
-            {
-                // W.r.t. specificity of above condition: Users are much more likely to inject their own
-                // IUrlHelperFactory than to create a class implementing IUrlHelper (or a sub-interface) and inject
-                // that. But the second scenario is supported. (Note the class must implement ICanHasViewContext.)
-                valueAccessor = _urlHelperAccessor;
-            }
-            else if (property.PropertyType == typeof(IJsonHelper))
-            {
-                valueAccessor = _jsonHelperAccessor;
-            }
-            else if (property.PropertyType == typeof(DiagnosticSource))
+            var type = page.GetType();
+            Type modelType;
+            if (!_modelTypeLookup.TryGetValue(type, out modelType))
```

Comment:
Calculating the `ModelType` (for VDD) is different in a Razor view and Razor pages.